### PR TITLE
Ownerbranch

### PR DESCRIPTION
**Pull request by tobezhanabi**

### CODE REVIEW

File: src/questions/owner_borrow.yaml [hunk 1]
```diff
@@ -240,3 +240,751 @@
...
+      key: B
+    - content: "Validator"
+      key: c
+    - content: "Static memory"
+      key: D
+
+- uuid: 1edea51c-a9f7-4ef7-a243-c96f29daff52
+  type: SingleChoice
+  content: In Rust, memory is managed through a system of ownership and rules that the compiler checks. If any of the rules are violated, what will happen to the program?
+  hint: No memory, no calculation
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: Hint and explanation don't match the question

File: src/questions/owner_borrow.yaml [hunk 2]
```diff
@@ -240,3 +240,751 @@
...
+  subTopics:
+    - ownership
+  difficultyLevel: low
+  choices:
+    - content: "0"
+      key: A
+    - content: "2"
+      key: B
+    - content: "1"
+      key: c
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: make `c` Upper Case

Comment (RobinNagpal) [reply src/questions/owner_borrow.yaml]: Do it for all the questions

File: src/questions/owner_borrow.yaml [hunk 3]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "3"
+      key: D
+
+- uuid: a86625dd-2eba-4808-87c0-30cde8cf7c2d
+  type: SingleChoice
+  content: |
+    Complete the sentence
+    When the owner goes out of _____, the value will be ______.
+  hint: NoHint
+  explanation: "Owners have right and if they abandon the right it get dropped"
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: Not a proper sentence "Owners have right and if they abandon the right it get dropped"

File: src/questions/owner_borrow.yaml [hunk 4]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "1"
+      key: c
+    - content: "3"
+      key: D
+
+- uuid: a86625dd-2eba-4808-87c0-30cde8cf7c2d
+  type: SingleChoice
+  content: |
+    Complete the sentence
+    When the owner goes out of _____, the value will be ______.
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]:
```suggestion
    
    Each value in Rust has an owner. There can only be one owner at a time. When the owner goes out of _____, the value will be ______.
    
```

File: src/questions/owner_borrow.yaml [hunk 5]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "a reference is guaranteed to point to a valid value of a particular type for the life of that reference."
+      key: B
+    - content: "new concept for many programmers, it does take some time to get used to"
+      key: c
+    - content: "the range within a program for which an item is valid"
+      key: D
+
+- uuid: 3661c275-8559-40e4-b5dd-ac138e3c5be9
+  type: SingleChoice
+  content: The action of creating a reference to a variable is called ____
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: This sentence is not complete in itself.

`Borrowing is a concept that allows us to borrow a value, use it and once we are done, return it back to its owner in the same state it was when we borrowed it. This means that borrowing does not change the original value.`

Comment (RobinNagpal) [reply src/questions/owner_borrow.yaml]: Here is the full sentence from https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html

`We call the action of creating a reference borrowing. As in real life, if a person owns something, you can borrow it from them. When you’re done, you have to give it back. You don’t own it.`

File: src/questions/owner_borrow.yaml [hunk 6]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "Heaping"
+      key: c
+    - content: "Borrowing"
+      key: D
+
+- uuid: 434bc0ff-02a9-4287-a469-e74c3b673f70
+  type: SingleChoice
+  content: What is a Scope?
+  hint: noHint
+  explanation: ""
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: Explanation is Mandator. The code will break if we dont add it

File: src/questions/owner_borrow.yaml [hunk 7]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "the range within a program"
+      key: c
+    - content: "None of the above"
+      key: D
+
+- uuid: 896206c4-09f5-4c46-8cfd-6f5597df0c02
+  type: SingleChoice
+  content: When your code calls a function, the values passed into the function are pushed into
+  hint: not a first in first out
+  explanation: "function are pushed in and dont have to leave first"
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: `function are pushed in and dont have to leave first`. this doesn't explain the question

File: src/questions/owner_borrow.yaml [hunk 8]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "the range within a program for which an item is invalid"
+      key: B
+    - content: "the range within a program"
+      key: c
+    - content: "None of the above"
+      key: D
+
+- uuid: 896206c4-09f5-4c46-8cfd-6f5597df0c02
+  type: SingleChoice
+  content: When your code calls a function, the values passed into the function are pushed into
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: The question is not clear. The values are already saved on to the heap and then function uses stack

```
Well, when a function gets called, some memory gets allocated for all of its local variables and some other information. This is called a ‘stack frame’, and for the purpose of this tutorial, we’re going to ignore the extra information and only consider the local variables we’re allocating. So in this case, when main() is run, we’ll allocate a single 32-bit integer for our stack frame. This is automatically handled for you, as you can see; we didn’t have to write any special Rust code or anything.

When the function exits, its stack frame gets deallocated. This happens automatically as well.

```

File: src/questions/owner_borrow.yaml [hunk 9]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "Memory"
+      key: c
+    - content: "calldata"
+      key: D
+
+- uuid: 762de2f4-3e2f-42a3-ba93-770ab7a833dc
+  type: SingleChoice
+  content: The function's local variables get pushed onto
+  hint: First in First Out
+  explanation: "pushed accordingly"
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: `pushed accordingly` is not a correct explanation

File: src/questions/owner_borrow.yaml [hunk 10]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "Heap"
+      key: B
+    - content: "Memory"
+      key: c
+    - content: "calldata"
+      key: D
+
+- uuid: 762de2f4-3e2f-42a3-ba93-770ab7a833dc
+  type: SingleChoice
+  content: The function's local variables get pushed onto
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: Also its not that they are pushed to stack.   New variables are saved in stack. 

Both mean essentially same, but the question can be reframed better

File: src/questions/owner_borrow.yaml [hunk 11]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "Calldata"
+      key: B
+    - content: "Function"
+      key: c
+    - content: "Stack"
+      key: D
+
+- uuid: aa969d59-a963-442d-b13b-d83ee2f65f28
+  type: SingleChoice
+  content: When the function is over, those values get_______ the stack.
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml:452]: `popped out` and `popped off` are both correct.

File: src/questions/owner_borrow.yaml [hunk 12]
```diff
@@ -240,3 +240,751 @@
...
+  hint: correct term for memory allocation
+  explanation: "The right term is pop out"
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: We can explain it in a better and detail way

----

Well, when a function gets called, some memory gets allocated for all of its local variables and some other information. This is called a ‘stack frame’, and for the purpose of this tutorial, we’re going to ignore the extra information and only consider the local variables we’re allocating. So in this case, when main() is run, we’ll allocate a single 32-bit integer for our stack frame. This is automatically handled for you, as you can see; we didn’t have to write any special Rust code or anything.

When the function exits, its stack frame gets deallocated. This happens automatically as well.

File: src/questions/owner_borrow.yaml [hunk 13]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "popped out"
+      key: B
+    - content: "popped off"
+      key: c
+    - content: "pushed out"
+      key: D
+
+- uuid: 2fe49404-ebe4-4183-a243-510840dbfdcc
+  type: SingleChoice
+  content: A problem that ownership addresses have is .
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml:475]: Ownership addresses multi threading problem

File: src/questions/owner_borrow.yaml [hunk 14]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "line 2"
+      key: B
+    - content: "line 3"
+      key: c
+    - content: "line 4"
+      key: D
+
+- uuid: 4531ae97-640b-4f5c-9c59-4d00633a3f9c
+  type: SingleChoice
+  content: What is the special annotation in Rust that we can place on types that are stored on the stack, as integers
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml]: The question and explanation are not framed correctly

File: src/questions/owner_borrow.yaml [hunk 15]
```diff
@@ -240,3 +240,751 @@
...
+    - content: "runtimes"
+      key: B
+    - content: "borrow checker"
+      key: c
+    - content: "FIFO"
+      key: D
+
+- uuid: e6b050d1-22fe-4375-bdcb-91a90f6c2385
+  type: SingleChoice
+  content: What kind of error does violating an ownership rules gives
```
Comment (RobinNagpal) [inline src/questions/owner_borrow.yaml:982]: Do we get compilation or panic error?  

Also add an explanation

Comment (tobezhanabi) [reply src/questions/owner_borrow.yaml:982]: its panic https://www.makeuseof.com/rust-ownership/#:~:text=unless%20it's%20borrowed.-,The%20Ownership%20Rules,variable%20owns%20its%20initialized%20value.

File: src/questions/owner_borrow.yaml (hunk 13 above)
Comment (tobezhanabi) [reply src/questions/owner_borrow.yaml:475]: i dont understand

File: src/details/owner_borrow.yaml [hunk 16]
```diff
@@ -1,5 +1,405 @@
+- title: Understanding ownership
+  shortTitle: ownership
+  key: ownership
+  details: |
+    - Ownership is Rust’s most unique feature, and it enables Rust to make memory safety guarantees without needing a garbage collector.
+    - Rust’s central feature is ownership. All programs have to manage the way they use a computer’s memory while running.
+    - Some languages have garbage collection that constantly looks for no longer used memory as the program runs; in other languages, the programmer must explicitly allocate and free the memory.
```
Comment (RobinNagpal) [inline src/details/owner_borrow.yaml]: Copying the sentences is not acceptable at all

`All programs have to manage the way they use a computer’s memory while running. Some languages have garbage collection that regularly looks for no-longer used memory as the program runs; in other languages, the programmer must explicitly allocate and free the memory. Rust uses a third approach: memory is managed through a system of ownership with a set of rules that the compiler checks. `
https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html#what-is-ownership

File: src/details/owner_borrow.yaml [hunk 17]
```diff
@@ -1,5 +1,405 @@
...
+    - Rust uses a third approach: memory is managed through a system of ownership with a set of rules that the compiler checks at compile time.
+    - None of the ownership features slow down your program while it’s running.
+    - The Rust compiler checks if a program obeys the ownership rules at compile time. If the program follows these rules, it can run. If it doesn’t, the compiler refuses to produce an executable.
+    - Rust verifies the ownership rules using the borrow checker. The borrow checker verifies the ownership model and decides if a value in memory (stack or heap) is out of scope or not.
+    - If a value is out of its scope, it’s not accessible to other parts of the program unless it’s borrowed.
+    - When you understand ownership, you’ll have a solid foundation for understanding the features that make Rust unique.
+    - The Ownership Rules
+      - In Rust, every variable owns the value it’s initialized with, and there can only be one owner.
```
Comment (RobinNagpal) [inline src/details/owner_borrow.yaml]: This is a copy paste https://www.makeuseof.com/rust-ownership/#:~:text=Rust%20verifies%20the%20ownership%20rules,the%20program%20unless%20it's%20borrowed.

**Pull request closed without merging**